### PR TITLE
[attributetable] Fix sorting with geometry expression

### LINKED ISSUE #57338: attribute table row sorting doesn't work
### What is the bug or the crash?

The sorting of the rows of the attribute table does not work if we use a geometric function as an expression (for example like $area)

### Steps to reproduce the issue

1. I load the Reg01012024_g_WGS84 layer in QGIS;
2. I open the attributes table;
3. I position the mouse on the COD_REG attribute header;
4. right-click and select 'sort';
5. in the window that opens, I activate it and type $area;
6. I click OK
7. nothing happens (even if you activate ascending sort in the window)

https://github.com/qgis/QGIS/assets/7631137/0f33bf65-427a-49f7-9483-96e98fbe435e

## OSGeo4W QGIS 3.34.6-Prizren

Versione di QGIS 3.34.6-Prizren
Revisione codice QGIS
623828f5
Versione Qt 5.15.13
Versione Python 3.12.3
Versione GDAL/OGR 3.8.5
Versione PROJ 9.4.0
Versione database del Registro EPSG v11.004 (2024-02-24)
Versione GEOS 3.12.1-CAPI-1.18.1
Versione SQLite 3.45.1
Versione PDAL 2.6.3
Versione client PostgreSQL 16.2
Versione SpatiaLite 5.1.0
Versione QWT 6.2.0
Versione QScintilla2 2.14.1
Versione SO Windows 11 Version 2009

Plugins Python attivi
DataPlotly 4.1.0
eurostat_downloader 0.2.1
felt 2.0.1
kmltools 3.1.33
mapswipetool_plugin 1.2
nominatim_locator_filter 0.3.2
quick_map_services 0.19.34
valuetool 3.0.19
db_manager 0.1.20
grassprovider 2.12.99
MetaSearch 0.3.6
processing 2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

data for testing:

[Reg01012023_g.zip](https://github.com/qgis/QGIS/files/15203663/Reg01012023_g.zip)

### PR DESCRIPTION
**Pull request by elpaso**
Fix #57338

### CODE REVIEW

File: src/app/qgsattributetabledialog.cpp [hunk 1]
```diff
@@ -217,16 +217,18 @@ QgsAttributeTableDialog::QgsAttributeTableDialog( QgsVectorLayer *layer, QgsAttr
   {
     request.setFilterExpression( filterExpression );
   }
-  if ( !needsGeom )
-    request.setFlags( Qgis::FeatureRequestFlag::NoGeometry );
 
+  if ( !needsGeom && ( !mLayer || !QgsExpression( mLayer->attributeTableConfig().sortExpression() ).needsGeometry() ) )
+  {
+    request.setFlags( Qgis::FeatureRequestFlag::NoGeometry );
+  }
```
Comment (nyalldawson) [inline src/app/qgsattributetabledialog.cpp:224]: I'd find this slightly more readable:

```suggestion
  // If sort expression requires geometry, we'll need to fetch it
  needsGeom |= mLayer && QgsExpression( mLayer->attributeTableConfig().sortExpression() ).needsGeometry();  
  if ( !needsGeom )
    request.setFlags( Qgis::FeatureRequestFlag::NoGeometry );
```

File: src/gui/attributetable/qgsdualview.cpp [hunk 2]
```diff
@@ -336,16 +339,13 @@ void QgsDualView::setFilterMode( QgsAttributeTableFilterModel::FilterMode filter
...
 
   const bool requiresTableReload = ( request.filterType() != Qgis::Qgis::FeatureRequestFilterType::NoFilter || request.spatialFilterType() != Qgis::SpatialFilterType::NoFilter ) // previous request was subset
                                    || ( needsGeometry && request.flags() & Qgis::FeatureRequestFlag::NoGeometry ) // no geometry for last request
                                    || ( mMasterModel->rowCount() == 0 ); // no features
 
-  if ( !needsGeometry )
-    request.setFlags( request.flags() | Qgis::FeatureRequestFlag::NoGeometry );
-  else
-    request.setFlags( request.flags() & ~( static_cast< int >( Qgis::FeatureRequestFlag::NoGeometry ) ) );
+  request.setFlags( request.flags().setFlag( Qgis::FeatureRequestFlag::NoGeometry, !needsGeometry ) );
```
Comment (nyalldawson) [inline src/gui/attributetable/qgsdualview.cpp:348]: Not necessary here, but I do think having a `QgsFeatureRequest::setFlag( Qgis::FeatureRequestFlag flag, bool enabled )` would be useful!

**Review by nyalldawson — APPROVED** [review]

File: src/app/qgsattributetabledialog.cpp (hunk 1 above)
Comment (agiudiceandrea) [reply src/app/qgsattributetabledialog.cpp:224]: Or `needsGeom = needsGeom || ( mLayer && QgsExpression( mLayer->attributeTableConfig().sortExpression() ).needsGeometry() );` to short-circuit when needsGeom is already True?

Comment (nyalldawson) [reply src/app/qgsattributetabledialog.cpp:224]: That's exactly the same as `|=`

**Comment by nyalldawson** [comment]
@elpaso I'm fine with that!

File: src/app/qgsattributetabledialog.cpp (hunk 1 above)
Comment (agiudiceandrea) [reply src/app/qgsattributetabledialog.cpp]: Sorry, I supposed `x |= y;` doesn't short-circuit like `x = x || y();` so `y` was always evaluated even if `x` is already true.

**Pull request merged**